HIP/hip-980.md

```diff
@@ -0,0 +1,120 @@
+---
+hip: 980
+title: Enhance mirror node rest APIs with timestamp query parameter
+author: Ivan Kavaldzhiev <ivan.kavaldzhiev@limechain.tech>
+working-group: Steven Sheehy <@steven-sheehy>, Nana Essilfie-Conduah <nana@swirldslabs.com>
+type: Standards Track
+category: Mirror
+needs-council-approval: Yes
+status: In review
+created: 2024-06-03
+requested-by: Solo Labs
+discussions-to: https://github.com/hashgraph/hedera-improvement-proposal/discussions/981
+updated: 2024-06-25
+---
+
+## Abstract
+
+This HIP enhances part of the mirror node REST APIs to support querying by timestamp. This will allow users to query for the historical state of an entity on the network.
+
+## Motivation
+
+To enhance user experience and allow users or developers to query the state of an entity a specific point in time we will add a timestamp query parameter to all of the entity-based REST APIs that currently don't have one.
+This change will be especially valuable for building the foundation of the so-called shadow fork functionality, where a user or developer can use a local mirror node instance but run it against a production environment state.
+Having this support, users or developers will have the ability to use public state data to simulate and troubleshoot existing issues or replay historical transactions using local mirror node instance against production state.
+
+## Rationale
+
+Users may want to run historical transactions on a local mirror node instance using public state data. To make this possible, some of the existing REST APIs should be enhanced with a timestamp query parameter, so that historical public data can be fetched and used
+for the simulations.
+
+## User stories
+
+- As a user of mirror node, I want to be able to query for the historical state of accounts, tokens, and other network entities.
+- As a user of mirror node, I want to be able to run locally an `eth_call`, `eth_estimateGas` or `eth_debugTraceTransaction` call that fails on a production environment against the corresponding public state.
```

Comment:
Similar to above I would break this out into a user story per API noted

Reply:
Done